doc/api/diagnostics_channel.md

```diff
@@ -0,0 +1,189 @@
+# Diagnostics Channel
+
+<!--introduced_in=REPLACEME-->
+
+> Stability: 1 - Experimental
+
+<!-- source_link=lib/diagnostics_channel.js -->
+
+The `diagnostics_channel` module provides an API to report diagnostic data
+via named channels. It can be accessed using:
+
+```js
+const diagnostics_channel = require('diagnostics_channel');
+```
+
+## Public API
+
+### Overview
+
+Following is a simple overview of the public API.
+
+```js
+const diagnostics_channel = require('diagnostics_channel');
+
+// Subscribe to a channel by name
+diagnostics_channel.subscribe('my-channel', (name, data) => {
+  // Received data
+});
```

Comment:
Looking at this, my mind keeps going back to why can't we just use `EventEmitter` here. I know there's the subscribe variation taking the object, but a modification to `EventEmitter` could potentially handle that in a semver-minor way... e.g. `ee.onAny(/^\s+-channel/, (event, ...args) => { /* ... */ })`

Reply:
It's conceptually similar. This is quite a bit faster than an event emitter though due to eliminating the lookups at publish time and being able to make publish a noop until something subscribes to a channel. Event emitter would need a rewrite to have the separate channel objects to be capable of matching the speed of this.

The intention is for this to be able to be used fairly liberally and subscribers could pick their specific relevant subset of things to listen to, so the overhead needs to be as low as possible when nothing is listening.

Reply:
I get that, I just really don't like having yet another events-like API that is similar but isn't quite the same.

Reply:
The alternative is we continue patching the world, breaking apps, and incurring unpredictable and often significant overhead. Given that the vast majority of apps use some APM or other, I think having an extra events API is by far the lesser of two evils.

Not to mention we _already_ have history of introducing APIs with significant overlap, like `worker_threads` while `cluster` is a thing, `URL` while `url.parse(...)` exists, the ongoing effort around a new http client with undici, `AsyncLocalStorage` doing what `async_hooks` was designed for in the first place but failed to adequately deliver.

Also, event emitter _could_ be rebuilt on top of this if we're really so concerned about that small bit of duplication. The reverse is not possible while achieving at all acceptable performance for the intended use.

Reply:
Also, just touching on the regex event emitter idea specifically: regex is quite expensive and I don't see a clear path on how to take that out of band from the publish path using the event emitter API. The API I have here allows it only because I was able to take the channel name lookup out of the publish path through having a separate publisher object that is constructed ahead of time and therefore binding the subscribers at creation time and not at publish time. Event emitters don't have this split between channel creation publish.

Reply:
But the intent of `URL` is to replace `url.parse()`. Likewise, the intent of `undici` is to ultimately replace the existing http client, and I wouldn't say that it's accurate to say that `AsyncLocalStorage` is doing what `async_hooks` "was designed for in the first place" as they really serve two distinct purposes as different levels. I'm not sure what exactly to suggest here, however, as I'm just providing feedback as requested. I'd be much happier if a reasonable way of using the existing APIs could be found.

Reply:
As would I, but the existing APIs just don't have the performance for this use case. This PR is about 10-30x faster than event emitters in my measurements. If I start putting emitter code into the key places we want, like http server code, the event emitter version makes Node.js _significantly_ slower while the code in this PR barely registers in my measurements.

If you have any concrete suggestions on how to adapt existing APIs to achieve this, I'm happy to consider it, but I've been working on APM products for Node.js for nearly a decade now and have spent a month now writing and benchmarking this code, and much longer planning for it, focusing heavily on performance. I just haven't been able to come up with a way to achieve this with event emitters without absolutely _destroying_ the performance of Node.js.

It's a fairly small amount of code, and I'm sure the numerous APM vendors with Node.js products would happily lend some time to help maintaining this if it meant they could have a safer and faster way to provide the data they need for their products, so I'm really not concerned about the similarity to the existing event emitter code.